storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectListOffsetsAction.java

```diff
@@ -19,33 +19,50 @@
 import org.apache.kafka.clients.admin.ListOffsetsResult;
 import org.apache.kafka.clients.admin.OffsetSpec;
 import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.storage.internals.epoch.LeaderEpochFileCache;
 import org.apache.kafka.storage.internals.log.EpochEntry;
 import org.apache.kafka.tiered.storage.TieredStorageTestAction;
 import org.apache.kafka.tiered.storage.TieredStorageTestContext;
 
 import java.io.PrintStream;
 import java.util.Collections;
+import java.util.Optional;
 import java.util.concurrent.ExecutionException;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 public final class ExpectListOffsetsAction implements TieredStorageTestAction {
 
+    private final int brokerId;
     private final TopicPartition partition;
     private final OffsetSpec spec;
-    private final EpochEntry expected;
+    private EpochEntry expected;
 
-    public ExpectListOffsetsAction(TopicPartition partition,
+    public ExpectListOffsetsAction(int brokerId,
+                                   TopicPartition partition,
                                    OffsetSpec spec,
                                    EpochEntry expected) {
+        this.brokerId = brokerId;
         this.partition = partition;
         this.spec = spec;
         this.expected = expected;
     }
 
     @Override
     public void doExecute(TieredStorageTestContext context) throws InterruptedException, ExecutionException {
+        if (expected.epoch == Integer.MAX_VALUE) {
+            // The leaderEpoch for the partition gets bumped/incremented when the partition gets reassigned (or)
+            // a new leader gets elected. But, the leaderEpoch value is non-deterministic between the ZK and Kraft mode.
+            // So, reading the leader-epoch value from the broker's LeaderEpochFileCache for the given offset.
+            Optional<LeaderEpochFileCache> leaderEpochFileCache = context.leaderEpochFileCache(brokerId, partition);
+            assertTrue(leaderEpochFileCache.isPresent());
+            leaderEpochFileCache.get().epochForOffset(expected.startOffset).ifPresent(epoch ->
+                expected = new EpochEntry(epoch, expected.startOffset)
```

Comment:
Could you explain more why the leaderEpoch is non-deterministic? This works though, it makes the testSpec is not readable when verifying listOffset.

Reply:
I tried running the ListOffsetsTest for both ZK and Kraft modes:

1. In ZK mode, the leader-epoch gets bumped during reassignment (0 -> 1 -> 2) and leader-election (2 -> 3)
2. In Kraft mode, the leader-epoch gets bumped only for leader-election (0 -> 1) and not for reassignment

So, the test fails for one mode and succeeds in another due to non-deterministic leader-epoch bump between Kraft and ZK modes. Since ZK mode is deprecated now, I'll run this test only for Kraft mode.